sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/inputFileBlock.scala

```diff
@@ -39,7 +40,7 @@ case class InputFileName() extends LeafExpression with Nondeterministic {
 
   override def nullable: Boolean = false
 
-  override def dataType: DataType = StringType
+  override def dataType: DataType = SQLConf.get.defaultStringType
```

Comment:
qq: do we need to change all expressions that return `StringType`?

Reply:
I would say yes. Suppose that session level default collation is "UNICODE", then all string function outputs should return `StringType("UNICODE")`, because `StringType` (which is `StringType(0)`) would be rather arbitrary in that setting.

In addition, string outputs can then be used as input for other functions, etc. and we do need to preserve types properly

Reply:
I agree it's dangerous, but there appears to be no obvious way around it. Here's how we're battling that for now:

1. manually adding collation support and testing collation-awareness for each function (we've already covered around 50% of Spark expressions)
2. working on RQG tests for collation, which should hopefully catch anything we missed along the way (this is also in progress within the collation effort)


And here's two scenarios of what happens if we unfortunately don't implement this for an expression:

Reply:
### Example 1 (incorrect behaviour with respect to collation)

User wants to check whether input file name (which is a string) begins with "abc"
```
SELECT startsWith(input_file_name(), "abc");
```

- suppose the user has set the session level default collation, let's say "UNICODE"
- `input_file_name` expression returns StringType(0), let's say it returns "ABCDE"
- `startsWith` expression takes (StringTypeAnyCollation, StringTypeAnyCollation)
- manually specified Literal "abc" will have the default collation - here: "UNICODE"

so in this example query, `startsWith` expression would execute with respect to first parameter collation (in this case 0 - which is "UTF8_BINARY"), which is not expected here because the session level default collation is "UNICODE"

Reply:
### Example 2 (query analysis fail in CollationTypeCasts)

User wants to check whether input file name (which is a string) begins with result from another function call -let's call it `foo()` for the sake of this example
```
SELECT startsWith(input_file_name(), foo());
```

- suppose the user has set the session level default collation, let's say "UNICODE"
- `input_file_name` expression which doesn't yet support collations, returns type StringType("UTF8_BINARY")
- `foo` expression which **does** support collations, returns type StringType("UNICODE")
- `startsWith` expression takes (StringTypeAnyCollation, StringTypeAnyCollation)

so in this example query, `startsWith` expression would fail due to a mismatch between parameter input types (2 different implicit collations, where neither has priority over the other)

Reply:
TLDR: Conclusion is that we will eventually need to support all expressions, and while there is a risk we'll miss some - there doesn't seem to be a way around it

Reply:
ok makes sense. Thanks for two  concrete examples to demonstrate two possible error case.  I see we have a config to control collation support:

```
  lazy val COLLATION_ENABLED =
    buildConf("spark.sql.collation.enabled")
      .doc("Collations feature is under development and its use should be done under this" +
        "feature flag.")
      .version("4.0.0")
      .booleanConf
      .createWithDefault(Utils.isTesting)
```
As long as this is by default false for real workload before we are confident to turn on the flag, then it is fine.